pkg/submitter/submitter.go

```diff
@@ -0,0 +1,569 @@
+		}),
+		submitFailCount: promauto.NewCounter(prometheus.CounterOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_fails_total",
+			Help:        "The total number of failed submission",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		}),
+		submitProfit: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_profit",
+			Help:        "The current submit profit in percents",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+		submitCost: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_cost",
+			Help:        "The current submit cost in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+		submitReward: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_reward",
+			Help:        "The current reward in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+	}
+
+	return submitter, submitter.resultCh, nil
+}
+
+func (s *Submitter) Start() error {
+	for {
+		select {
+		case <-s.ctx.Done():
+			if s.resultCh != nil {
+				close(s.resultCh)
+			}
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			level.Info(s.logger).Log("msg", "submitter shutdown complete")
+			return s.ctx.Err()
+		case result := <-s.resultCh:
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			var ctx context.Context
+			ctx, s.lastSubmitCncl = context.WithCancel(s.ctx)
+
+			level.Info(s.logger).Log("msg", "received a solution",
+				"addr", result.Work.PublicAddr,
+				"challenge", fmt.Sprintf("%x", result.Work.Challenge),
+				"solution", result.Nonce,
+				"difficulty", result.Work.Challenge.Difficulty,
+				"requestIDs", fmt.Sprintf("%+v", result.Work.Challenge.RequestIDs),
+			)
+			s.blockUntilTimeToSubmit(ctx)
+			s.handleSubmit(ctx, result)
+		}
+	}
+}
+
+func (s *Submitter) Stop() {
+	s.close()
+}
+
+func (s *Submitter) blockUntilTimeToSubmit(newChallengeReplace context.Context) {
+	var (
+		lastSubmit time.Duration
+		timestamp  *time.Time
+		err        error
+	)
+	for {
+		select {
+		case <-newChallengeReplace.Done(): // The context was canceled from the main loop because new work arrived.
+		default:
+		}
+		lastSubmit, timestamp, err = s.lastSubmit()
+		if err != nil {
+			level.Debug(s.logger).Log("msg", "checking last submit time", "err", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		break
+	}
+	if lastSubmit < s.cfg.Mine.MinSubmitPeriod.Duration {
+		level.Debug(s.logger).Log("msg", "min transaction submit threshold hasn't passed", "minSubmitPeriod", s.cfg.Mine.MinSubmitPeriod, "lastSubmit", lastSubmit)
```

Comment:
This can also be an info now since it will show only once and not repeat every few seconds.
```suggestion
		level.Info(s.logger).Log("msg", "min transaction submit threshold hasn't passed", "minSubmitPeriod", s.cfg.Mine.MinSubmitPeriod, "lastSubmit", lastSubmit)
```